### ISSUE TITLE
ENH cache datasets

### ISSUE THREAD

**Issue by jolars** [benchopt/benchopt#468]
I think it makes sense to provide our own caching mechanism for datasets in benchopt. In many benchmarks we currently rely on libsvmdata to do the caching, but I think it makes more sense to do this in benchopt, especially because new contributors that want to add datasets from other sources need to then write their own caching functions, which in the end means there will be multiple cache locations that there is no control over through benchopt.

I can provide a PR if you think this sounds like a good idea.

**Comment by jolars** [comment]
> Just to make sure: we already have a caching mechanism with BaseDataset._get_data that loads the data only once per benchmark. But you're talking about once and for all caching so that data is downloaded only once by the user, for example ?

Yes, sorry, that's right. I mean that downloaded datasets are cached. 

> where do we store it? how does the user specify a path if there is already a caching mechanism, eg sklearn data of libsvm data?

I don't think the user should specify the path. We should select a path in a platform-independent way, for instance using appdirs. The name could then be something like `/var/cache/benchopt/benchmark_lasso/data/url` for the url data set in the benchmark_lasso repo.

> how do we handle datasets with the same name between benchmarks?

We would need to store duplicates. Anything else would just be too complicated and prone to fail. Yes, that's a definite downside. Maybe that one is large enough to make us stay away from doing this.

**Comment by mathurinm** [comment]
Btw, if the concern comes from the Breheny datasets, @theoguyard is adding them to libsvmdata this week : https://github.com/mathurinm/libsvmdata/pull/36

**Comment by jolars** [comment]
That's nice, but no, not really. I'm more concerned about making it easy for new collaborators to add datasets.
